rfcs/001 - EC2 schema.md

```diff
@@ -0,0 +1,113 @@
+
+The definition of a solid base now will imply that new features added to this module will not translate into breaking changes to the model.
+
+Taking a look at the AWS EC2 console options, we would still need to add the following to reach EC2 module completeness:
+
+- Instance type - Availability zone relationship
+- Launch templates
+- Spot instances
+- Dedicated hosts
+- Saving plans
+- Reserved instances
+- Scheduled instances
+- Capacity reservations
+- Auto Scaling Groups
+- AMIs management
+- EBS volume snapshots and life cycle
+- Elastic IP integration
+- Placement groups
+- Network interfaces
+
+Almost all the features listed will need a direct integration with the `instance` table, needing to add at least eight more columns to the model. Those columns are only for integration, but the `instance` model still needs more properties to be defined to be complete. Some of those missing properties are: instance auto-recovery, shutdown behaviour, stop - hibernate behaviour, termination protection, stop protection, detailed cloudWatch monitoring, elastic GPU, credit specification, tenancy, RAM disk ID, kernel ID or Metadata accessible (I might be missing others).
+
+The current instance module has ten columns. Adding the ones representing integration with other features will almost double its size.
+
+## Proposal
+
+The `instance` table is the core of the `aws_ec2` module and is the one that can grow more in complexity. The proposal is to keep adding the necessary columns to the `instance` table. It is better to have a unique table to look for instances than increase the complexity of the model, creating other tables and adding to the users the need to understand our model first. Eventually, we can add an `aws_ec2_simplified` module that can expose the basics to create an instance and leave the `aws_ec2` for advanced users.
+
+We can divide the feature list as follow:
+
+The direct implication in the `instance` table:
+
+- Launch templates
+- Spot instances
+- Dedicated hosts
+- Capacity reservations
+- Auto Scaling Groups
+- Elastic IP integration
+- Placement groups
+- Network interfaces
+
+No implication in the `instance` table:
+
+- Instance type - Availability zone relationship
+- AMIs management
+- EBS volume snapshots and life cycle
+
+The `Saving plans`, `Reserved instances` or `Scheduled instances` are not included in the list above since they seem to be budgeting solutions that need interaction with the user rather than infrastructure-related modules' features that could have any implication in the model.
+
+Let's start with the features without impact on the `instance` table. These features will be part of the `aws_ec2` module with their mappers. The Instance type table by availability zones will be a table with a composite primary key. It technically has a minimum impact in the `instance` table that will be a special FK to the composite primary key, but no UX model impact.
```

Comment:
how would eventually going multi-region affect this part of the proposal? would it affect other parts of this proposal?

Reply:
Hmm good point. It will depend on how we implement that. For instance, assuming it will be a `region` column in all tables, this will work, we just need to add the new column to the composite key and thats all. At first glance I do not see other parts of this proposal affected by the multi-region change.